---
title: "lasso_model"
author: "Kate Weaver"
date: "1/10/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, load_packages, warning=FALSE, message=FALSE}
library(SummarizedExperiment)
library(rtracklayer)
library(tidyverse)
library(data.table)
library(readxl)
library(janitor)
library(ggplot2)
library(AnnotationDbi)
library(EnsDb.Hsapiens.v86)
library(glmnet)
library(glmmLasso)
library(edgeR)
```

importing SummarizedExperiment Data and transforming to raw counts data
```{r load_data_se}
load("~/mccoyLab/collabs/doubleseq_2021/summarized_experiment/create_summarized_experiment_allNov2021.Rdata")
counts <- as.matrix(assays(seAll)$counts)
for (col in 1:ncol(counts)){
  colnames(counts)[col]<-strsplit(sub("Aligned.sortedByCoord.out.bam", "", colnames(counts)[col]), "_")[[1]][1]
}
```

```{r, load_andsubsetby_sets}
#subset to just training data/samples
embryoID_by_set <- read.csv("~/mccoyLab/collabs/doubleseq_2021/data_split/embryo_bySet_full_kw_20211217.csv", col.names=c("embryoID", "set"))
train_embryoIDs = embryoID_by_set$embryoID[embryoID_by_set$set == "train"]
```

```{r load_andsubsetby_gencode}
#subset to genes with matchind gencode ensembl gene IDs on chr1-22
file_gencode <- "~/genomes/hg38_genome/gencode.v34.annotation.gtf"

gtf <- rtracklayer::import(file_gencode) %>% 
  as.data.frame() %>% 
  dplyr::filter(type == "gene") %>% 
  dplyr::select(gene_id, seqnames, width, gene_name) %>% 
  dplyr::rename(ensembl_gene_id = gene_id) %>% 
  dplyr::rename(chromosome_name = seqnames) %>% 
  dplyr::rename(gene_name = gene_name) %>% 
  dplyr::rename(length = width)

gene_table <- gtf[match(rownames(counts), gtf$ensembl_gene_id),]
gene_table <- gene_table[gene_table$chromosome_name %in% paste0("chr", 1:22),]
counts <- counts[gene_table$ensembl_gene_id,]
```

```{r load_andmatchby_meta}
#Importing metadata to differentiate between pregnant and not pregnant and other main covariates
meta <- read.csv("~/mccoyLab/collabs/doubleseq_2021/tidied_meta/tidied_meta_CREATE_kw_20211217.csv", row.names = 1) %>% 
  as.data.frame() %>% 
  mutate(across(c("AOD", "GC", "Infertility_type", "Previous_pregnancy", "Past_surgical_hist", "Pregnant", "Ongoing_pregnancy", "Final_outcome", "Embryo_grade_at_freezing", "Interpretation", "cDNA_RT_Date", "Library_Prep_Date", "Sequencing_Date", "Study_Participant_ID"), as.factor)) %>%
  mutate(across(c("InfD_SSM_GC", "InfD_Egg_factor", "InfD_MF", "InfD_Uterine_factor", "InfD_TF", "InfD_RPL", "InfD_RIF", "InfD_Unexplained", "PMdH_none", "PMdH_vasculitis", "PMdH_immune", "PMdH_stress_hormones"), as.factor))

#subsetting embryos to include only ones with sequencing data 
remove <- setdiff(train_embryoIDs, colnames(counts))
if (length(which(rownames(meta) %in%  remove)) > 0){
  train_embryoIDs <- train_embryoIDs[-which(train_embryoIDs %in% remove)]
}

#subsetting meta and counts data to just the training rows            
meta <- meta[train_embryoIDs,]
counts <- as.matrix(t(as.data.frame(t(counts))[train_embryoIDs,])) #57640 genes

#need to sort so name order is the same for setting up design for DESeq experiment
counts_order <- order(colnames(counts))
meta_order <- order(rownames(meta))

countdata <- counts[, counts_order]
coldata <- meta[meta_order, ] #107 samples
```

```{r filter_expression}
#Use more stringent filtering
dgeFullData <- DGEList(countdata, group=as.factor(coldata$Study_Participant_ID))
#normalize counts by TMM
TMMFullData <- calcNormFactors(dgeFullData, method="TMM")
TMMCounts <- as.matrix(TMMFullData$counts)
countsCleaned <- TMMCounts[rowSums(TMMCounts >= 6) > (ncol(TMMCounts)* .2),]

#TPM Calculation
calc_tpm <- function(x, gene.length) {
  x <- as.matrix(x)
  len.norm.lib.size <- colSums(x / gene.length)
  return((t(t(x) / len.norm.lib.size) * 1e06)/ gene.length)
}

#creates a matrix with calculated TPM values for each sample from TMM normalized counts and the gene lengths
rawTPMvals <- calc_tpm(TMMFullData, gene.length = gene_table$length)
cleanedTPMVals <- rawTPMvals[rowSums(rawTPMvals > 0.1) > (ncol(rawTPMvals)*.2),]
cleanCountsDf <- as.data.frame(countsCleaned)
cleanTPMdf <- as.data.frame(cleanedTPMVals)

countdata <- cbind(cleanTPMdf[rownames(cleanCountsDf),]) #22955 genes
```


select only relevant metadata columns
```{r subsetmeta}
columns_to_remove <- c("Ongoing_pregnancy", "Final_outcome",
                       "Time_to_OPU_hr", "Time_to_stripping_hr", "Time_to_ICSI_hr", "Time_to_Biopsy_hr", "AMH",
                       "Sildenafil")
to_remove <- which(colnames(meta) %in% columns_to_remove)


meta_covs <- meta[,-to_remove]
meta_covs$Pregnant=meta_covs$Pregnant==1
```

select countdata genes from any of difexpression analyses
```{r downselect_genes}
design0_res <- read.csv("~/mccoyLab/collabs/doubleseq_2021/dif_expression_results/results_trainSPA_onlyBatch.qval.csv")
design3_res <- read.csv("~/mccoyLab/collabs/doubleseq_2021/dif_expression_results/results_trainSPA_dupCor_withBatch.qval.csv")
featselect_res <- read.csv("~/mccoyLab/collabs/doubleseq_2021/dif_expression_results/results_trainSPA_featselect_withBatch.qval.csv")
genes_oi <- full_join(full_join(featselect_res, design3_res), design0_res)
countdata <- countdata[genes_oi$X,] #only 64 genes
```

make a full dataframe with sample as row and covariate variable or gene as column
```{r onedf}
df <- merge(meta_covs, as.data.frame(t(countdata)), by=0) %>%
  `rownames<-`(df$Row.names) %>% as.data.frame() %>% 
  dplyr::select(-Row.names) %>%
  dplyr::mutate_at(vars(starts_with("ENSG")), as.numeric) #107 samples; 117 variables
```

```{r, lassomodel}
fix_form <- reformulate(setdiff(colnames(df), c("Pregnant", "Study_Participant_ID", "cDNA_RT_Date", "Library_Prep_Date", "Sequencing_Date","Transferring_physician", "Transfer_catheter_used")), response="Pregnant")
fix_form
```

```{r, lassomodel_pt2}
rnd_form <- list("Study_Participant_ID"=~1, "cDNA_RT_Date"=~1, "Library_Prep_Date"=~1,"Sequencing_Date"=~1, "Transferring_physician"=~1, "Transfer_catheter_used"=~1)
lassomod <- glmmLasso(fix = Pregnant ~ as.factor(AOD) + as.factor(GC) + Oocyte_Age + Uterus_Age + Sperm_Age + BMI + 
    as.factor(Infertility_type) + as.factor(Previous_pregnancy) + as.factor(Past_surgical_hist) + 
    FSH_D3 + AFC + LH + E2_Day_2_3 + as.factor(Type_of_trigger) + Total_Gn + 
    E2_on_trigger + Num_eggs + Num_MII + Num_2PN + Num_Blasts + 
    TMC + as.factor(Embryo_grade_at_freezing) + as.factor(Interpretation) + as.factor(Natural_cycle) + 
    as.factor(Medication) + lining_thickness_mm + as.factor(Prednisone) + as.factor(Fragmin) + 
    as.factor(EG) + as.factor(HCG) + as.factor(IL) + as.factor(GCSF) + as.factor(Metformin) + as.factor(PRP) + as.factor(InfD_SSM_GC) + as.factor(InfD_Egg_factor) + 
    as.factor(InfD_MF) + as.factor(InfD_Uterine_factor) + as.factor(InfD_TF) + as.factor(InfD_RPL) + as.factor(InfD_RIF) + 
    as.factor(InfD_Unexplained) + as.factor(PMdH_none) + as.factor(PMdH_vasculitis) + as.factor(PMdH_immune) + 
   as.factor(PMdH_stress_hormones) + ENSG00000139908.15 + ENSG00000198526.7 + 
    ENSG00000260876.6 + ENSG00000274979.1 + ENSG00000235523.1 + 
    ENSG00000158220.14 + ENSG00000197496.6 + ENSG00000143297.19 + 
    ENSG00000135355.4 + ENSG00000276754.1 + ENSG00000139531.13 + 
    ENSG00000197651.4 + ENSG00000159882.13 + ENSG00000287462.1 + 
    ENSG00000285696.1 + ENSG00000243742.5 + ENSG00000163689.20 + 
    ENSG00000184992.13 + ENSG00000228889.7 + ENSG00000167419.11 + 
    ENSG00000234277.3 + ENSG00000286963.1 + ENSG00000139438.6 + 
    ENSG00000251244.2 + ENSG00000170236.16 + ENSG00000164167.12 + 
    ENSG00000186723.4 + ENSG00000170949.17 + ENSG00000235713.1 + 
    ENSG00000140265.12 + ENSG00000151445.16 + ENSG00000171243.8 + 
    ENSG00000214894.6 + ENSG00000164338.10 + ENSG00000186417.14 + 
    ENSG00000233208.6 + ENSG00000260698.1 + ENSG00000253550.1 + 
    ENSG00000187720.14 + ENSG00000163746.11 + ENSG00000254180.2 + 
    ENSG00000198440.9 + ENSG00000279315.1 + ENSG00000118292.9 + 
    ENSG00000117281.16 + ENSG00000138496.16 + ENSG00000283069.1 + 
    ENSG00000131059.12 + ENSG00000230299.1 + ENSG00000223522.2 + 
    ENSG00000206043.7 + ENSG00000280739.3 + ENSG00000164707.15 + 
    ENSG00000259683.1 + ENSG00000118271.11 + ENSG00000090612.22 + 
    ENSG00000075239.14 + ENSG00000112984.12 + ENSG00000065308.5 + 
    ENSG00000235275.3 + ENSG00000120875.9 + ENSG00000160570.14 + 
    ENSG00000160570.14.1 + ENSG00000054965.10
, 
                      rnd = rnd_form,
                      lambda = 2,
                      df,
                      family = binomial(link="logit"))
```
