---
title: "infer_sex"
author: "Kate Weaver"
date: "1/7/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

```{r load_packages, message=FALSE, warning=FALSE}
library(SummarizedExperiment)
library(rtracklayer)
library(tidyverse)
library(data.table)
library(readxl)
library(janitor)
library(AnnotationDbi)
library(EnsDb.Hsapiens.v86)
library(ggplot2)
library(edgeR)
```

importing SummarizedExperiment Data and transforming to raw counts data
```{r load_data_se}
load("~/mccoyLab/collabs/doubleseq_2021/summarized_experiment/create_summarized_experiment_allNov2021.Rdata")
counts <- as.matrix(assays(seAll)$counts)
for (col in 1:ncol(counts)){
  colnames(counts)[col]<-strsplit(sub("Aligned.sortedByCoord.out.bam", "", colnames(counts)[col]), "_")[[1]][1]
}
```


```{r TPM_function}
#TPM Calculation
calc_tpm <- function(x, gene.length) {
  x <- as.matrix(x)
  len.norm.lib.size <- colSums(x / gene.length)
  return((t(t(x) / len.norm.lib.size) * 1e06)/ gene.length)
}
```


```{r load_andsubsetby_gencode}
#subset to genes with matching gencode ensembl gene IDs
file_gencode <- "~/genomes/hg38_genome/gencode.v34.annotation.gtf"

gtf <- rtracklayer::import(file_gencode) %>% 
  as.data.frame() %>% 
  dplyr::filter(type == "gene") %>% 
  dplyr::select(gene_id, seqnames, width, gene_name) %>% 
  dplyr::rename(ensembl_gene_id = gene_id) %>% 
  dplyr::rename(chromosome_name = seqnames) %>% 
  dplyr::rename(length = width)

gene_table <- gtf[match(rownames(counts), gtf$ensembl_gene_id),]
counts <- counts[gene_table$ensembl_gene_id,]
```


```{r compute_tpm}
dgeFullData <- DGEList(counts)
TMMFullData <- calcNormFactors(dgeFullData, method="TMM")
rawTPM <- calc_tpm(TMMFullData, gene.length=gene_table$length)
```


```{r subsetby_chrom}
#on chrY
gene_table <- gene_table[gene_table$chromosome_name == "chrY",]
rawTPM <- rawTPM[gene_table$ensembl_gene_id,]
```

```{r three_genes}
g1 <- "DDX3Y"
g2 <- "RPS4Y1"
g3 <- "EIF1AY"
roi <- which(gene_table$gene_name == g1 | gene_table$gene_name == g2 | gene_table$gene_name == g3)
rawTPM_threegene <- rawTPM[roi,]
gene_table_threegene <- gene_table[roi,]
```

```{r sumacrossthree}
sampleSum <- colSums(rawTPM_threegene)
```

```{r plotit}
toPlot <- data.frame(summed_tpm = unlist(sampleSum), sample = names(sampleSum))
ggplot(toPlot, aes(x=summed_tpm)) + geom_histogram(binwidth = 1) +
  theme_bw() + theme(panel.grid = element_blank(), panel.background = element_blank()) + ggtitle("TPM Summed Accross DDX3Y, RPS4Y1, and EIF1AY")
```
```{r}
names(sampleSum)[sampleSum >= 10]
```